### PR TITLE
Accept masked arrays for Data XXX creation

### LINKED ISSUE #346: unexpected behavior with masked arrays
Sherpa version 4.9.0, Python 2.7.12,  NumPy 1.11.2 (also seen on NumPy 1.12.1)
This issue was discovered when fitting a masked 2D array but is easily reproducible with the first few lines of the example notebook:

Create a masked numpy array:
```
x = np.arange(-5, 5.1)
y = x*x + 23.2 + np.random.normal(size=x.size)
old_y = x*x + 23.2 + np.random.normal(size=x.size)
y = np.ma.masked_array(old_y,mask=y<35)
e = np.ones(x.size)

sherpa.load_arrays("mydata", x, y, e)
sherpa.plot_data("mydata")
print(y.mask)

```
The mask appears set but the data points still appear.
If you multiply the masked array by 1.0 (or as I noticed this, by taking the sqrt of an array) the behavior changes:

```
y = 1.0*np.ma.masked_array(old_y,mask=y<35)
sherpa.load_arrays("mydata", x, y, e)
sherpa.plot_data("mydata")
print(y.mask)
```

Screenshots:

screen shot 2017-03-31 at 4 29 50 pm

This still produces unexpected behavior, as these new zero values aren't just plotted, they got included in the fit further down the notebook. 
### Work around
Forcing masked values in y to np.nan gives the expected behavior, e.g.
`y = np.ma.filled(1.0*np.ma.masked_array(old_y,mask=y<35),fill_value=np.nan)` then the masked y data points are neither plotted or fitted.

### LINKED ISSUE #695: Initializing generid data silently drops mask
When I make a `Data1D` or `Data2D` dataset from a numpy masked array, Sherpa silently drops the mask and uses the values "behind the mask".
```
import numpy as np

from sherpa.data import Data1D, Data2D
from sherpa.fit import Fit
from sherpa.models import Const1D, Const2D

arr1d = np.ma.array([1.,2.,90.,3.,1.], mask=[False, False, True, False, False])
arr2d = np.ma.array([[1,1.5], [10., 1.2]], mask = [[False, False], [True, False]])

d1 = Data1D('d1', np.array([1,2,3,4,5]), arr1d)
d2 = Data2D('d2', np.array([1,1,2,2]), np.array([1,2,1,2]), arr2d.ravel())
print(d1.mask)

c1 = Const1D('c1')
fit1 = Fit(d1, c1)
print(fit1.fit())
```
The Fit results is (some non-relevant lines removed from output):
```
succeeded      = True
parnames       = ('c1.c0',)
parvals        = (2.8467372929128665,)
statval        = 69.91988835701966
numpoints      = 5
dof            = 4
qval           = 2.359979141585598e-14
rstat          = 17.479972089254915
```
Clearly, the fit is terrible (just look at qval) and the fitted values is too large. `numpoints` shows that the fit used all five datapoints, including the one that should have been masked.
However, I can manually set the mask (**note that I actually need to set the opposite of the mask** because numpy uses `True` to indicate masked values and Sherpa uses `True~ to indicate valid values and `False~ for masked values):
```
d1.mask = ~arr1d.mask
fit1 = Fit(d1, c1)
print(fit1.fit())
```
Now, the fit works better and uses only 4 unmasked datapoints:
```
succeeded      = True
parnames       = ('c1.c0',)
parvals        = (1.5945294293559473,)
statval        = 0.3833606897991587
numpoints      = 4
dof            = 3
qval           = 0.9436579269604303
rstat          = 0.12778689659971956
```

The same works for 2D datasets...
I think its very dangerous to accept masked arrays on input and silently drop the mask. That's even more confusing since there is no obvious way to pass in a mask. I had to discover the `data.mask` attribute and then figure out that the definition is exactly opposite to numpy. Very unfortunate choice!

So, what could be done? I suggest:

- Add a parameters `mask` to the initializers for the Data classes. That's analogous to `numpy.array.ma(data=..., mask=...)`. When people try to pass in a masked array and it does not work as expected, looking at the parameters for initialization is what I did first.
- Document that mask convention **WITH BIG WARNINGS**.
- Accept `np.ma.array` on input and automatically convert the mask. Alternatively, issue warning if any datatype with a `.mask` attribute is passed in and say "Dropping mask. Use mask= parameter to explicitly set mask".
- It may be to late to change Sherpa's True/False convention for masks. But it's very bad (TM) that this is opposite to numpy, in particular since Sherpa uses numpy for all numerics. In the long run, we should investigate if the Sherpa data classes could use `numpy.ma.array` as their underlying datatype instead of carrying a separate mask. I have not looked through the code, but I imagine that Sherpa must implement a lot of functions to deal with masked data itself, that we could get for free from numpy.ma, starting by simple things like sum, mean and average of an array, taking into account the mask, to matplotlib integration (matplotlib knows not to plot masked values if given a masked array).

### PR DESCRIPTION
**Pull request by hamogu**
Accept masked arrays for Data XXX creation
    
This change implements changes to use the mask of numpy arrays
when initializing DataXXX. It warns sensibly for non-numpy objects with
a mask in the OO interface when duck-typing, but not in the session
interface, because that whitelists only specific data types anyway.
Thus, this reverses the warnings added to the docs in #698.
    
fixes #695
fixes #346

### CODE REVIEW

**Comment by hamogu** [comment]
Note that he test failures stem from"
```
PackagesNotFoundError: The following packages are not available from current channels:

  - libgfortran=3.0
```
That's clearly not due to this PR but an unrelated CI issue. I will try to restart the jobs tomorrow and try again.

**Comment by hamogu** [comment]
#794 might fix the CI, so that this can be restarted.

File: sherpa/data.py [hunk 1]
```diff
@@ -52,6 +52,24 @@ def _check(array):
     return array
 
 
+def _check_nomask(array):
```
Comment (DougBurke) [inline sherpa/data.py:55]: When I see a message like "Dropping mask attribute" I think I'd expect for the attribute to be removed in some way. I'm not sure that's the behavior here - the .mask attribute may stay but it is just ignored. Or am I missing something?

I'm not 100% convinced that `del array.mask` would be a sensible thing here, but I wanted to understand the warning message.

File: sherpa/data.py [hunk 2]
```diff
@@ -381,6 +399,11 @@ def mask(self):
     def mask(self, val):
         if (val is True) or (val is False):
             self._mask = val
+        # if val is of type np.bool_ and True, it was failed the previous test because
+        # "is True" compares with Python "True" singelton.
+        # Yet, we do not want to allow arbitrary values that evaluate as True.
+        elif (val is numpy.ma.nomask) or (numpy.isscalar(val) and val and isinstance(val, numpy.bool_)):
+            self._mask = True
```
Comment (DougBurke) [inline sherpa/data.py:406]: I would have expected something like

```
elif val is numpy.ma.mask:
    salf._mask = True
elif numpy.isscalar(val) and isinstance(val, numy.bool_):
    self._mask = val  # do we need to convert from numpy bool to bool?
elif ...
```

because don't we want `_mask` to be False if `val` is `numpy.bool_(0)`? Or does that get handled by another check in the chain?

File: sherpa/data.py [hunk 3]
```diff
@@ -464,17 +487,13 @@ class Data(NoNewAttributesAfterInit, BaseData):
...
-        to-be-ignored value.
-
         Parameters
         ----------
         name : basestring
             name of this dataset
         indep: tuple of array_like
-            the tuple of independent arrays
+            the tuple of independent arrays. If this is an numpy masked array,
+            the mask will used.
```
Comment (DougBurke) [inline sherpa/data.py:496]: what does 'the mask will be used' mean? Is it "the masked array will be used" or "the unmasked values" or ...? I don't use the masked array stuff so I don't know the expected terminology.

Actually, this is the independent array: isn't the mask ignored in this case? Actually, it depends on _init_data_space does, which isn't immediately obvious.

File: sherpa/tests/test_data.py [hunk 4]
```diff
@@ -977,3 +1014,128 @@ def test_data2d_int_eval_model_to_fit(array_sizes_fixture):
...
+    args[posy] = numpy.ma.array(args[posy])
+    session.load_arrays(*args)
+    new_data = session.get_data(data.name)
+    # Sherpa's way of saying "mask is not set"
+    assert new_data.mask is True
+    assert len(new_data.get_dep(filter=True)) == len(args[posy].flatten())
+
+# https://github.com/sherpa/sherpa/issues/346
+@pytest.mark.parametrize('Session', [Session, AstroSession])
+def test_regression_346(Session):
```
Comment (DougBurke) [inline sherpa/tests/test_data.py:1133]: Not issue #346, but what happens if you try to create a Data object (either directly or via load_arrays) where all points are masked out? It looks like `Data1D` at least supports this, so I imagine it'll result in something like the following, but wanted to check:

```
In [20]: from sherpa.data import Data1D

In [21]: d = Data1D('x', [], [])
/home/dburke/sherpa/sherpa-master/sherpa/data.py:49: UserWarning: Converting array [] to numpy array
  warnings.warn("Converting array {} to numpy array".format(array))

In [22]: print(d)
name      = x
x         = Float64[0]
y         = Float64[0]
staterror = None
syserror  = None
```

File: sherpa/data.py [hunk 5]
```diff
@@ -464,17 +487,13 @@ class Data(NoNewAttributesAfterInit, BaseData):
 
     def __init__(self, name, indep, y, staterror=None, syserror=None):
         """
```
Comment (DougBurke) [inline sherpa/data.py:489]: I think we probably want some description added to the `Data` docstring to describe the handling of masked arrays.

File: sherpa/data.py (hunk 1 above)
Comment (hamogu) [reply sherpa/data.py:55]: I don't want to `del` anything, because we don't want to change the original array that the user had. I've reworded the messages, you'll see that on my next push.

**Comment by hamogu** [comment]
A general thought that will help me clarify docstrings and warning messages: I suggest to accept masks on dependent arrays only. Independent and dependent arrays are input to sherpa separately, so if we don't take care of it in some way, the user could input an x where some values are masked and a y where some other values are masked. What does that even mean? Also, for binned data, we input `xlo` and `xhi`. What about bins where the lower edge is masked, but the upper edge is not. That does not make any sense.

So, I decided that only masks that are defined on the dependent array will be accepted and used to set the filter. Having any mask defined on the independent array will simply raise a warning and continue with the raw values as if the mask never existed (the current behavior for independent arrays with a mask attribute, except for an added warning). 

My thinking is that x defines some way of measurement, e.g. the energy of a spectral bin, the time of data taken, or the input voltage of a diode. `y` is then the measured quantity (e.g. the number of photons, the flux in a time series, or the light output from the diode). `x` is never invalid in that case (i.e. in principle we would make a measurement at any energy, ant time or any voltage), but `y` might be invalid (e.g. we don't have data at some energy, it was cloudy at a specific night, or the diode burned out). 

Of course, these examples only go so far (e.g. there are no negative energies, so one could a argue the in this case the independent variable should be masked), but I think we need to make a choice here to keep code and user interface understandable, so I suggest: **Masks on independent arrays are ignored with a warning, masks on dependent arrays are used to set_filter (if the True/False convention is clear, i.e. numpy masked arrays).**

After all, this is for convenience and to follow the principle of least surprise (if I put in a masked array, the mask does not just disappear silently as it currently does) - a user can always put in raw (unmasked) x and y array and set_filter after at their hearts content if they want something more complicated.

@DougBurke : Do you agree?

**Comment by anetasie** [comment]
@hamogu I agree. Masking on dependent arrays make sense. On the other hand one may want to filter the dependent arrays on the values of Y, so say, ignore the data with `Y<10` for example.  That should be possible with the user function that setups the filter, which internally sets a mask on the dependent array.

**Comment by hamogu** [comment]
@anetasie I'm not sure I understand where you are going with your comment. This is just about being able to use numpy masked arrays in `load_array` and using the mask in the original array to set the filter. Once the data is read in to sherpa, you can do whatever you can do now, i.e. set the filter in any way you like with `set_filter`.

**Comment by hamogu** [comment]
@DougBurke Does this address your concerns?

File: sherpa/data.py (hunk 2 above)
Comment (DougBurke) [reply sherpa/data.py:416]: I guess the next question is whether we have any tests of this to check our logic makes sense? I know our existing test set doesn't do a good job of exploring the edge cases, so it would be interesting to know if there's any conditions that aren't evaluated.

File: sherpa/tests/test_data.py (hunk 4 above)
Comment (DougBurke) [reply sherpa/tests/test_data.py:1158]: @hamogu - do you know what happens in this case ? I didn't see any obvious test changes that exercised this, but I could well have missed it.

File: sherpa/data.py (hunk 2 above)
Comment (hamogu) [reply sherpa/data.py:416]: Well, I've added a LOT of tests (about a dozen, but they are all parameterized, (e.g. with and without mask) so I've lost track of the exact number, but I believe it's > 50.

Comment (DougBurke) [reply sherpa/data.py:416]: I meant have you checked the HTML coverage output to see if anything isn't being covered? It does require running the tests with coverage turned on of course, but I think it should just be 'pytest --cov=sherpa` and then `coverage html` (I never remember the option to name the output directory, or indeed the exact commands you need to use).

Comment (DougBurke) [reply sherpa/data.py:416]: I've rebased the branch onto master (so I can build it with a half-way recent set of software) and the coverage for the following (with XSPEC/astropy/matplotlib installed and the test dataset) indicates that lines 415 and 419 were never True, so their code was never executed. Which is why there's fancy unicode in the output below

```
409    def mask(self, val): 
410        if (val is True) or (val is False): 
411            self._mask = val 
412        # if val is of type np.bool_ and True, it failed the previous test because 
413        # "is True" compares with Python "True" singelton. 
414        # Yet, we do not want to allow arbitrary values that evaluate as True. 
415        elif val is numpy.ma.nomask: 415 ↛ 416
416            self._mask = True 
417        elif numpy.isscalar(val) and isinstance(val, numpy.bool_): 
418            self._mask = bool(val) 
419        elif (val is None) or numpy.isscalar(val): 419 ↛ 420
420            raise DataErr('ismask') 
421        else: 
422            self._mask = numpy.asarray(val, numpy.bool_) 
```

**Comment by DougBurke** [comment]
@hamogu - do you want me to push the rebased version to this branch (or leave as is/do it yourself)?

File: sherpa/tests/test_data.py (hunk 4 above)
Comment (hamogu) [reply sherpa/tests/test_data.py:1158]: It behaves as it should. Your code above works exactly as before because it passes in lists, not numpy arrays and that behaviour is unchanged. If you use masked array, you can mask as much as you like. The masked datapoints are there and can be retried just as you can do after using `set_mask`. That behaviour is unchanged, too. All the changes is that instead of (assuming y is a numpy masked array)
```
d = Data1D('x', x, y)
d.set_mask(~y.mask)
```
you can now simply do `d=Data1D('x', x, y)`. It really just saves you that one line of typing. This PR is important not because it helps Sherpa do something it could not do before, but because is aligns the default to what users apparently expect (#346, #695) - and that expectation "Don't just ignore masks without even telling me you do that" is totally reasonable.

Full output from the tests @DougBurke asked for:
```
sherpa In [1]: from sherpa.data import Data1D 
          ...:                                                                                                       

sherpa In [2]: d = Data1D('x', [], [])                                                                               
/Users/guenther/anaconda3/envs/sherpaciao/lib/python3.7/site-packages/sherpa-4.12.1+301.gecfad5f4.dirty-py3.7-macosx-10.9-x86_64.egg/sherpa/data.py:49: UserWarning: Converting array [] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))

sherpa In [3]: print(d)                                                                                              
name      = x
x         = Float64[0]
y         = Float64[0]
staterror = None
syserror  = None

sherpa In [6]: xarr = numpy.ma.array([1,2,3], mask=[True, True, True])                                               

sherpa In [7]: yarr = [1, 2, 1]                                                                                      

sherpa In [10]: d = Data1D('x', xarr, yarr)                                                                          

sherpa In [11]: print(d)                                                                                             
name      = x
x         = Int64[3]
y         = Int64[3]
staterror = None
syserror  = None

sherpa In [12]: d.get_indep()                                                                                        
Out[12]: (array([1, 2, 1]),)

sherpa In [13]: d.get_indep(filter=True)                                                                             
Out[13]: (array([], dtype=int64),)
```

Comment (DougBurke) [reply sherpa/tests/test_data.py:1158]: Thanks. It's an edge case to check, but someone always finds them ;-)

Comment (DougBurke) [reply sherpa/tests/test_data.py:1158]: I'll leave it up to you to see if it needs an explicit test or not.

**Comment by DougBurke** [comment]
So I'm just learning hot to properly decode the codecov output: the 87.5% coverage of the code added in this PR could be better, (I've noticed that my current PRs are at this level too) but there are reasons it'll never get to 100% for this PR, such as us not having a test on travis that uses crates [this is something I've been thinking about for a while]

**Comment by hamogu** [comment]
In astropy, I usually find is more instructive to look at the differences, i.e. a new piece of code should not add untested lines (with some exceptions).
And while Travis probably should have a setup with crates, I don't think it's worth the effort to merge the codecov outputs together or run the codecov twice.

**Comment by DougBurke** [comment]
@hamogu - the codecov output is already the merge of 4 (I think) runs; it is done automatically for us.

File: sherpa/data.py (hunk 2 above)
Comment (DougBurke) [reply sherpa/data.py:416]: I think we want to make sure to have a test of 'val is numpy.ma.nomask' evaluating to True.

Comment (hamogu) [reply sherpa/data.py:416]: This line is only triggered when directly setting the mask (when the mask is inferred from the mask attribute of an array on initialization, then it's already handelded i nthe `_check_y` function and never gets to this point. Added test.

Comment (DougBurke) [reply sherpa/data.py:416]: Thanks,

**Comment by hamogu** [comment]
@dtnguyen2 : Ping! Just waiting for your review.

**Comment by dtnguyen2** [comment]
OK, looking at it now

**Comment by hamogu** [comment]
@dtnguyen2 The line you pointed out is not required for this change, thus I did not touch it. But, as you say, it's good to be consistent, so I just pushed the change.
